ivadomed/losses.py

```diff
@@ -1,16 +1,54 @@
 import torch
 import torch.nn as nn
-import torch.nn.functional as F
+
+
+class MultiClassDiceLoss(nn.Module):
+    # Inspired from https://arxiv.org/pdf/1802.10508.pdf
+    def forward(self, prediction, target):
+        smooth = 1.0
+        dice_per_class = 0
+        n_classes = prediction.shape[1]
+        for i in range(n_classes):
+            iflat = prediction[:, i, ].reshape(-1)
+            tflat = target[:, i, ].reshape(-1)
+            intersection = (iflat * tflat).sum()
+            dice_per_class += intersection / (iflat.sum() + tflat.sum() + smooth)
+        return -(2.0 * dice_per_class) / n_classes
+
+
+class CombinedDiceLoss(nn.Module):
+    """
+    :param classes_of_interest: list containing the index of a class which dice will be added to the loss.
+    :param multiclass_dice: boolean representing whether or not multiclass dice is included in the current loss.
+    :param dice: boolean representing whether or not dice is included in the current loss.
+    """
+
+    def __init__(self, class_of_interest, multi_class_dice, dice):
+        super().__init__()
+        self.class_of_interest = class_of_interest
+        self.multi_class_dice = multi_class_dice
+        self.dice = dice
+
+    def forward(self, prediction, target):
+        loss = torch.zeros(1).cuda()
+        if len(self.class_of_interest):
+            for idx in self.class_of_interest:
+                loss += - dice_loss(prediction[:, idx, ], target[:, idx, ])
+        if self.multi_class_dice:
+            loss += MultiClassDiceLoss()(prediction, target)
```

Comment:
Between these two cases: I can see only one difference: the `/ n_classes` in `MultiClassDiceLoss` --> is it correct?
Just wondering if we could combine all cases in only one loop (cf above comment).
--> This class would be the one to call for both mono and multi label tasks using the Dice Loss

Reply:
I think I understand what you mean! I optimized MultiClassDiceLoss and CombinedDiceLoss -> tell me if you had something else in mind